lib/wpscan/wp_target.rb

```diff
@@ -177,4 +178,24 @@ def upload_directory_listing_enabled?
   def include_directory_listing_enabled?
     directory_listing_enabled?(includes_dir_url)
   end
+
+  # Return the set of 404 error page hash for wp-content directories.
+  #
+  # This set is useful to check if a plugin or a theme is installed
+  # because 404 error pages for non-existent items can be different from
+  # target's default because of some plugin(i.e. IP Geo Block).
+  #
+  # @return [ Set<String> ]
+  def error_404_hash_set
+    unless @error_404_hash_set
+      hash_set = Set.new([site_404_hash])
+
+      non_existant_page = Digest::MD5.hexdigest(rand(999_999_999).to_s) + '.html'
+      [wp_content_dir, wp_plugins_dir, wp_themes_dir].each do |dir|
```

Comment:
All these directories contain blank index.php files for example. A lot of sites render a blank page when calling these directories so this would produce a lot of false positives. So we need some more checks in this place.

Also you are copying the `site_404_hash` method from above. Please use a single method for this so we do not duplicate code.

Reply:
Sorry, I don't understand why you think blank index page brings false-positive.

Suppose a target returns following responses against each URL,

```
- #{random-md5-digest}.html                       -> 404                 hash X
- /wp-content/plugins/                            -> 200                 hash Y
- /wp-content/plugins/#{random-md5-digest}.html   -> 200 (but not found) hash Z
```

response hashes for other non-existent URLs under the plugin directory are expected to be Z (or maybe X) and this PR intends to treat them as "not found".

Whether directory index is blank page or not (= hash Y's value) has nothing to do with item detection, does it?

Reply:
> Also you are copying the site_404_hash method from above. Please use a single method for this so we do not duplicate code.

`wp_conent_dir` method needs site's 404 page hash via `default_wp_content_dir_exists?` method, but `error_404_hash_set` cannot be used because it calls `wp_content_dir` and raises stackoverflow.
Therefore, `site_404_hash` method must be separated.

Reply:
The problem is we check for wp-themes, wp-plugins and so on several times and I think if we qualify this hast as a not found hash, this will break some stuff :/

Reply:
Also if one of the directories has directory listing enabled, this would be treated like a 404.

Reply:
@FireFart  I think the idea that referring the common hash set for any type of detection brings complexity and risk of breaking other parts.

It seems to be better that plugin's 404 hash would be used only for plugin detection (my first approach).
The same thing can be said about themes.

I've tried it on [another branch](https://github.com/wpscanteam/wpscan/compare/master...l3msh0:avoid-false-positive-detection-for-each-item-type). Please check it out.

>  If a plugin 404 gives hash x and target 404 a y, this would not fire if the plugin dir for some reason returns hash y (the hash from the target).

This problem remains but I don't think it is so important compared to the cost of introducing the hash set.

Reply:
@FireFart ping